### PR TITLE
Week8 Ex 2-5

### PR DESCRIPTION
**Pull request by tienwei**
Thanks @stevemao. Not sure if I made the Ex3 right? Please review.

### CODE REVIEW

**Comment by stevemao** [comment]
Try using `testCompany` and `testCompany2` to test and see what you get?

**Comment by tienwei** [comment]
I got 30 instead of 26 like yours XD

**Comment by stevemao** [comment]
Should be 26 if you count it manually.

**Comment by tienwei** [comment]
Fixed! Thanks @stevemao!

**Comment by tienwei** [comment]
image
Ex 5 result.

**Comment by stevemao** [comment]
Is is the result from company.txt?

**Comment by tienwei** [comment]
@stevemao Yes, is the result(fun score and names) correct?

**Comment by stevemao** [comment]
The right answer should be 33200

**Comment by tienwei** [comment]
Will review it again XD. Thanks!

**Comment by tienwei** [comment]
@stevemao finally fixed! 
image

File: cis194/week8/tien/src/Party.hs [hunk 1]
```diff
@@ -20,3 +22,27 @@ instance Monoid GuestList where
...
 moreFun = max
+
+-- exercie 2 --
+treeFold :: (a -> [b] -> b) -> Tree a -> b
+treeFold f (Node l []) = f l []
+treeFold f (Node l s) = f l $ treeFold f <$> s
+
+-- exercie 3 --
+nextLevel :: Employee -> [(GuestList, GuestList)] -> (GuestList, GuestList)
+nextLevel emp [] = (glCons emp mempty, mempty)
```
Comment (stevemao) [inline cis194/week8/tien/src/Party.hs]: This is probably redundant.

Comment (tienwei) [reply cis194/week8/tien/src/Party.hs]: Removed. Thanks!

File: cis194/week8/tien/app/Main.hs [hunk 2]
```diff
@@ -1,4 +1,14 @@
...
+
 main :: IO ()
-main = putStrLn "no use"
+main = do
+  dir <- getCurrentDirectory
+  handle <- openFile (dir ++ "/src/company.txt") ReadMode
+  contents <- hGetContents handle
+  putStrLn $ "Total Fun: " ++ (show . totalFun . read $ contents)
+  mapM_ putStrLn . totalEmpNameList . read $ contents
+  hClose handle
```
Comment (stevemao) [inline cis194/week8/tien/app/Main.hs]: I think we can still move some stuff out and only keep `IO` related things here.

File: cis194/week8/tien/app/Main.hs [hunk 3]
```diff
@@ -1,4 +1,14 @@
...
+import System.IO
+
 main :: IO ()
-main = putStrLn "no use"
+main = do
+  dir <- getCurrentDirectory
+  handle <- openFile (dir ++ "/src/company.txt") ReadMode
+  contents <- hGetContents handle
+  putStrLn $ "Total Fun: " ++ (show . totalFun . read $ contents)
+  mapM_ putStrLn . totalEmpNameList . read $ contents
```
Comment (stevemao) [inline cis194/week8/tien/app/Main.hs]: `read` is partial. We can add a simple error handling if company.txt is malformed.

File: cis194/week8/tien/src/Party.hs [hunk 4]
```diff
@@ -20,3 +22,26 @@ instance Monoid GuestList where
...
+-- exercie 2 --
+treeFold :: (a -> [b] -> b) -> Tree a -> b
+treeFold f (Node l []) = f l []
+treeFold f (Node l s) = f l $ treeFold f <$> s
+
+-- exercie 3 --
+nextLevel :: Employee -> [(GuestList, GuestList)] -> (GuestList, GuestList)
+nextLevel emp gls = (gl1, gl2)
+  where
+    gl1 = glCons emp $ foldr (<>) mempty $ snd <$> gls
```
Comment (shineli1984) [inline cis194/week8/tien/src/Party.hs]: `foldr (<>) mempty` seems like `fold` from `Data.Foldable`. It comes with `base`
```suggestion
    gl1 = glCons emp . fold . fmap snd $ gls
```

**Pull request merged**